### ISSUE TITLE
Implement dependency tracking

### ISSUE THREAD

**Issue by jharmison-redhat** [jharmison-redhat/openshift-devsecops#67]
As the workshop content continues to grow in complexity, we need a method for dependency tracking and dynamic dependency checking.

I envision a way where we could allow users to set their toggles appropriately and we don't deviate from that. If the deployer disables a role that another role depends on, it should fail with a descriptive message. I think using a YAML file that tracks individual dependencies per role could be parsed, and the roles that would be imported for a given variable matrix (`deploy_crw` for example) would determine if a role would not have been imported.

For example, changing `vars/devsecops.yml` to something like this:

```yaml
deploy_toggles:
  crw: no
  workshop: yes
# etc, etc
```

Would need the requisite changes to `roles/devsecops/tasks/main.yml` to support the new style, probably with a mechanism to combine the selected toggles with a default, for example:

```yaml
- name: Override default role toggles as user selected
  set_fact:
    combined_toggles: '{{ default_deploy_toggles|combine(deploy_toggles) }}'
````

Then a new task list could be pre-imported, for example `roles/devsecops/tasks/preflight.yml`, that would determine what roles would be imported for a given combined `deploy_toggles` dictionary. We could use a dictionary like this:

```yaml
toggle_to_role:
  crw: codeready-workspaces
  workshop: workshop-spawner
```

Along with an import like this:

```yaml
- import_tasks: '{{ role_path }}/../{{ toggle_to_role[item[0]] }}/prereq.yml'
  loop: '{{ combined_toggles|dict2items }}'
  when:
  - item[1]|bool
  - lookup('first_found', role_path + '/../' + toggle_to_role[item[0]] + '/prereq.yml', skip=yes) != []
```

Where, for example, `roles/workshop-spawner/prereq.yml` contained the following:

```yaml
---
- fail:
    msg: |
      The workshop spawner depends on CodeReady Workspaces.
      Please set `deploy_toggles.crw` to `true` in your `vars/devsecops.yml` file.
  when: not combined_toggles.crw
```

This simple example should be enough for the rest to make sense. We could also include more granular failures for things like toggles within roles, but as long as we're sticking with the monorepo, doing dependencies like this enables us to prevent unexpected behavior for the deployer (e.g. meta.dependencies pulling in things they thought they had disabled) while still maintaining a clear way to determine dependencies without wasting time at runtime.

**Comment by jharmison-redhat** [comment]
We also have an opportunity here to clean up our deployment toggles and role names, if it might be more appropriate to set them to values that are more related to one another.

**Comment by jharmison-redhat** [comment]
also *NOTE*: I didn't run any of this example code. I don't know if it works. I hope it's enough to get across what I'm trying to say via the idea before we actually implement anything :)

**Comment by ghost** [comment]
A concern with this approach is that it can quickly devolve into a mess b/c it permits declarations of children/dependencies as peers at the "deploy_toggles" level.
In the example provided, there will never be a deployment case where workshop  doesn't have a dependency on crw so having them as "toggle" element peers is a bit of a red herring. We are giving the folks using this for deployment the false sense that they can "chinese menu" the deployment and then correcting that assumption. 
I don't have a better approach to offer without adding a layering approach (possibly based on compositional complexity i.e. tiers of toggles)

**Comment by jharmison-redhat** [comment]
I think tiers of toggles could potentially solve both problems nicely, without alternative dependency tracking mechanisms. Disabling a tier disables all subordinate features in that tier, and that makes sense to the deployer.

**Comment by jharmison-redhat** [comment]
In the `devsecops` meta role this could be implemented with `- block:` task-level organization (which could potentially clutter the flow or make it hard to follow), or conditional imports (which would break up the meta-role's `main.yml` task list in a way that doesn't necessarily make sense).

I think both have trade-offs, but the approach either way sounds better than using a custom dependency resolution.

**Comment by jharmison-redhat** [comment]
See https://github.com/RedHatGov/devsecops-workshop-code/issues/7
